File: sdks/javascript/src/store.ts
/**
 * SQLite WAL durability layer — crash-safe event storage.
 *
 * Events are written here immediately after update(). Background flush
 * sends them to cloud. If the process crashes, pending events survive
 * and are retried on next connect().
 */

import Database from "better-sqlite3";

const SCHEMA = `
CREATE TABLE IF NOT EXISTS events (
    local_event_uuid TEXT PRIMARY KEY,
    bandit_id        INTEGER NOT NULL,
    arm_id           INTEGER NOT NULL,
    payload          TEXT NOT NULL,
    status           TEXT NOT NULL DEFAULT 'pending',
    created_at       REAL NOT NULL,
    human_reward     REAL,
    graded_at        REAL,
    s3_exported      INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS idx_events_status ON events(status);
`;

const MIGRATION_GRADING = [
  "ALTER TABLE events ADD COLUMN human_reward REAL",
  "ALTER TABLE events ADD COLUMN graded_at REAL",
  "ALTER TABLE events ADD COLUMN s3_exported INTEGER NOT NULL DEFAULT 0",
];

export interface EventPayload {
  local_event_uuid: string;
  bandit_id: number;
  arm_id: number;
  model_name: string;
  model_provider: string;
  bandit_name?: string;
  [key: string]: unknown;
}

export class EventStore {
  private db: Database.Database;
  private pushStmt: Database.Statement;
  private pendingStmt: Database.Statement;

  constructor(dbPath: string = ":memory:") {
    this.db = new Database(dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.pragma("busy_timeout = 5000");
    this.db.pragma("synchronous = NORMAL");
    this.db.exec(SCHEMA);
    this.migrate();

    // Pre-compile frequently-used statements
    this.pushStmt = this.db.prepare(
      `INSERT OR IGNORE INTO events
       (local_event_uuid, bandit_id, arm_id, payload, status, created_at)
       VALUES (?, ?, ?, ?, 'pending', ?)`,
    );
    this.pendingStmt = this.db.prepare(
      `SELECT local_event_uuid, payload FROM events WHERE status = 'pending'
       ORDER BY created_at ASC LIMIT ?`,
    );
  }

  /** Insert a pending event. */
  push(event: EventPayload): void {
    this.pushStmt.run(
      event.local_event_uuid,
      event.bandit_id,
      event.arm_id,
      JSON.stringify(event),
      Date.now() / 1000,
    );
  }

  /** Return up to `limit` pending events (oldest first). */
  pending(limit: number = 50): EventPayload[] {
    const rows = this.pendingStmt.all(limit) as {
      local_event_uuid: string;
      payload: string;
    }[];
    const results: EventPayload[] = [];
    const corrupt: string[] = [];
    for (const row of rows) {
      try {
        results.push(JSON.parse(row.payload) as EventPayload);
      } catch {
        console.warn(
          `[bandito] Discarding corrupt event ${row.local_event_uuid} from store`,
        );
        corrupt.push(row.local_event_uuid);
      }
    }
    if (corrupt.length > 0) {
      this.markFlushed(corrupt);
    }
    return results;
  }

  /** Mark events as successfully flushed to cloud. */
  markFlushed(uuids: string[]): void {
    if (uuids.length === 0) return;
    const placeholders = uuids.map(() => "?").join(",");
    this.db
      .prepare(
        `UPDATE events SET status = 'flushed'
         WHERE local_event_uuid IN (${placeholders})`,
      )
      .run(...uuids);
  }

  /** Record a human grade locally. */
  markGraded(uuid: string, reward: number): void {
    this.db
      .prepare(
        `UPDATE events SET human_reward = ?, graded_at = ?
         WHERE local_event_uuid = ?`,
      )
      .run(reward, Date.now() / 1000, uuid);
  }

  /** Return un-exported events as {event, ts} for S3 dump. */
  pendingS3(limit: number = 100): Array<{ event: Record<string, unknown>; ts: number }> {
    const rows = this.db
      .prepare(
        `SELECT payload, created_at FROM events WHERE s3_exported = 0
         ORDER BY created_at ASC LIMIT ?`,
      )
      .all(limit) as { payload: string; created_at: number }[];
    return rows.map((r) => ({ event: JSON.parse(r.payload) as Record<string, unknown>, ts: r.created_at }));
  }

  /** Mark events as successfully exported to S3. */
  markS3Exported(uuids: string[]): void {
    if (uuids.length === 0) return;
    const placeholders = uuids.map(() => "?").join(",");
    this.db
      .prepare(`UPDATE events SET s3_exported = 1 WHERE local_event_uuid IN (${placeholders})`)
      .run(...uuids);
  }

  /** Close the database connection. */
  close(): void {
    this.db.close();
  }

  private migrate(): void {
    for (const stmt of MIGRATION_GRADING) {
      try {
        this.db.exec(stmt);
      } catch {
        // Column already exists
      }
    }
  }
}

File: cli/src/store.rs
use anyhow::{Context, Result};
use rusqlite::Connection;
use std::collections::HashMap;
use std::path::{Path, PathBuf};

/// SQLite schema — mirrors the SDK's events table.
const SCHEMA: &str = "
CREATE TABLE IF NOT EXISTS events (
    local_event_uuid TEXT PRIMARY KEY,
    bandit_id        INTEGER NOT NULL,
    arm_id           INTEGER NOT NULL,
    payload          TEXT NOT NULL,
    status           TEXT NOT NULL DEFAULT 'pending',
    created_at       REAL NOT NULL,
    human_reward     REAL,
    graded_at        REAL,
    s3_exported      INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS idx_events_status ON events(status);
";

/// Read-only access to the SDK's local SQLite event store.
/// Used by the TUI to display query/response text without cloud round-trips.
pub struct EventStore {
    conn: Connection,
}

/// Text fields extracted from a local event payload.
pub struct EventText {
    pub query_text: Option<String>,
    pub response: Option<String>,
    pub system_prompt: Option<String>,
}

/// Event fields needed for judge calibration / augmentation.
pub struct JudgeEvent {
    pub uuid: String,
    pub arm_id: i64,
    pub query_text: Option<String>,
    pub response: Option<String>,
    #[allow(dead_code)]  // stored for future prompt enrichment
    pub system_prompt: Option<String>,
    pub human_reward: Option<f64>,
}

impl EventStore {
    pub fn open() -> Result<Option<Self>> {
        let path = Self::db_path()?;
        if !path.exists() {
            return Ok(None);
        }
        let conn = Connection::open_with_flags(
            &path,
            rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY | rusqlite::OpenFlags::SQLITE_OPEN_NO_MUTEX,
        )
        .with_context(|| format!("Failed to open {}", path.display()))?;
        Ok(Some(Self { conn }))
    }

    /// Create a writable SQLite at `path` (for S3 TUI hydration).
    pub fn create_ephemeral(path: &Path) -> Result<Self> {
        let conn = Connection::open(path)
            .with_context(|| format!("Failed to create ephemeral store at {}", path.display()))?;
        conn.execute_batch(SCHEMA)
            .context("Failed to initialize ephemeral store schema")?;
        Ok(Self { conn })
    }

    /// Insert a raw event payload downloaded from S3. No-op if UUID already present.
    pub fn insert_payload(&self, payload: &serde_json::Value, created_at: f64) -> Result<()> {
        let uuid = payload["local_event_uuid"].as_str().unwrap_or("").to_string();
        if uuid.is_empty() {
            return Ok(());
        }
        let bandit_id = payload["bandit_id"].as_i64().unwrap_or(0);
        let arm_id = payload["arm_id"].as_i64().unwrap_or(0);
        self.conn.execute(
            "INSERT OR IGNORE INTO events \
             (local_event_uuid, bandit_id, arm_id, payload, status, created_at) \
             VALUES (?1, ?2, ?3, ?4, 'flushed', ?5)",
            rusqlite::params![uuid, bandit_id, arm_id, payload.to_string(), created_at],
        )?;
        Ok(())
    }

    fn db_path() -> Result<PathBuf> {
        let home = dirs::home_dir().context("Could not determine home directory")?;
        Ok(home.join(".bandito").join("events.db"))
    }

    /// Fetch text fields for a list of event UUIDs.
    /// Returns a map from UUID to EventText.
    pub fn get_text(&self, uuids: &[String]) -> Result<HashMap<String, EventText>> {
        if uuids.is_empty() {
            return Ok(HashMap::new());
        }

        let placeholders: Vec<&str> = uuids.iter().map(|_| "?").collect();
        let sql = format!(
            "SELECT local_event_uuid, payload FROM events WHERE local_event_uuid IN ({})",
            placeholders.join(",")
        );

        let mut stmt = self.conn.prepare(&sql)?;
        let params: Vec<&dyn rusqlite::types::ToSql> =
            uuids.iter().map(|u| u as &dyn rusqlite::types::ToSql).collect();

        let mut result = HashMap::new();
        let rows = stmt.query_map(params.as_slice(), |row| {
            let uuid: String = row.get(0)?;
            let payload: String = row.get(1)?;
            Ok((uuid, payload))
        })?;

        for row in rows {
            let (uuid, payload) = row?;
            if let Ok(json) = serde_json::from_str::<serde_json::Value>(&payload) {
                let text = EventText {
                    query_text: json
                        .get("query_text")
                        .and_then(|v| v.as_str())
                        .map(String::from),
                    response: extract_response_text(&json),
                    system_prompt: json
                        .get("system_prompt")
                        .and_then(|v| v.as_str())
                        .map(String::from),
                };
                result.insert(uuid, text);
            }
        }

        Ok(result)
    }

    /// Fetch flushed events that have a human reward, stratified by arm.
    /// Returns up to `limit_per_arm` events per arm_id.
    pub fn get_graded_events(&self, bandit_id: i64, limit_per_arm: usize) -> Result<Vec<JudgeEvent>> {
        let mut stmt = self.conn.prepare(
            "SELECT local_event_uuid, arm_id, payload, human_reward \
             FROM events \
             WHERE bandit_id = ?1 AND human_reward IS NOT NULL AND status = 'flushed' \
             ORDER BY arm_id ASC, created_at DESC",
        )?;

        let rows = stmt.query_map([bandit_id], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, i64>(1)?,
                row.get::<_, String>(2)?,
                row.get::<_, f64>(3)?,
            ))
        })?;

        let mut counts: HashMap<i64, usize> = HashMap::new();
        let mut events = Vec::new();

        for row in rows {
            let (uuid, arm_id, payload, human_reward) = row?;
            let count = counts.entry(arm_id).or_insert(0);
            if *count >= limit_per_arm {
                continue;
            }
            *count += 1;

            let json: serde_json::Value =
                serde_json::from_str(&payload).unwrap_or(serde_json::Value::Null);
            let query_text = json.get("query_text").and_then(|v| v.as_str()).map(String::from);
            let response = extract_response_text(&json);
            let system_prompt = json.get("system_prompt").and_then(|v| v.as_str()).map(String::from);

            events.push(JudgeEvent {
                uuid,
                arm_id,
                query_text,
                response,
                system_prompt,
                human_reward: Some(human_reward),
            });
        }

        Ok(events)
    }

    /// Fetch flushed events with no human reward, stratified by arm.
    /// Returns up to `limit_per_arm` events per arm_id.
    pub fn get_ungraded_events(&self, bandit_id: i64, limit_per_arm: usize) -> Result<Vec<JudgeEvent>> {
        let mut stmt = self.conn.prepare(
            "SELECT local_event_uuid, arm_id, payload \
             FROM events \
             WHERE bandit_id = ?1 AND human_reward IS NULL AND status = 'flushed' \
             ORDER BY arm_id ASC, created_at DESC",
        )?;

        let rows = stmt.query_map([bandit_id], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, i64>(1)?,
                row.get::<_, String>(2)?,
            ))
        })?;

        let mut counts: HashMap<i64, usize> = HashMap::new();
        let mut events = Vec::new();

        for row in rows {
            let (uuid, arm_id, payload) = row?;
            let count = counts.entry(arm_id).or_insert(0);
            if *count >= limit_per_arm {
                continue;
            }
            *count += 1;

            let json: serde_json::Value =
                serde_json::from_str(&payload).unwrap_or(serde_json::Value::Null);
            let query_text = json.get("query_text").and_then(|v| v.as_str()).map(String::from);
            let response = extract_response_text(&json);
            let system_prompt = json.get("system_prompt").and_then(|v| v.as_str()).map(String::from);

            events.push(JudgeEvent {
                uuid,
                arm_id,
                query_text,
                response,
                system_prompt,
                human_reward: None,
            });
        }

        Ok(events)
    }
}

/// Extract response text from event payload.
/// Response can be a string or a dict with a "response" key.
fn extract_response_text(json: &serde_json::Value) -> Option<String> {
    match json.get("response") {
        Some(serde_json::Value::String(s)) => Some(s.clone()),
        Some(serde_json::Value::Object(obj)) => obj
            .get("response")
            .and_then(|v| v.as_str())
            .map(String::from),
        _ => None,
    }
}
